### PR TITLE
[Serve] log warning to reconfigure `max_ongoing_requests` if `max_batch_size` is less than `max_ongoing_requests`

### LINKED ISSUE #43288: [Serve] Dynamic Request Batching not sending max_batch_size
### What happened + What you expected to happen

When using @serve.batch with max_batch_size of e.g. 280, it's not waiting for 280 requests before the batch_wait_timeout_s finishes. 
I tried even set the batch_wait_timeout_s to a super large value e.g. 5 seconds, still not working. 

### Versions / Dependencies

ray[serve]==2.9.2

### Reproduction script

In a deployment class for model (note: there's only 1 replica for this class):
```
@serve.batch(max_batch_size=256, batch_wait_timeout_s=5)
async def __call__(self, inputs: list[any]) -> list[any]:
    result = self.model.batch_predict(inputs)
    self.logger.info(f"length called: {len(inputs)}")
    return result
```

If I call the above method with below example scripts:
This is in the parent deployment class which serves http traffic.
```
inference_tasks = [self.model.remote(prompt) for prompt in prompts] # prompts with length 140 
inference_results = await asyncio.gather(*inference_tasks)
```

The logs show:
```
length called: 101
length called: 39
```

Does the dynamic batching depends on other parameters? like GPU memory?

### Issue Severity

organization productivity impaired

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

Our existing request rate limiting is not factor in the batch size. When using a large `max_batch_size` (> `max_ongoing_requests`), the batch won't process the batch up to `max_batch_size` even when `batch_wait_timeout_s` is set large enough to collect requests. This PR look for each and every methods on the user code to find the max of `max_batch_size` on all methods and log a warning to suggest user to reconfigure `max_ongoing_requests` in order to reach `max_batch_size` limit. 

## Related issue number

Closes https://github.com/ray-project/ray/issues/43288

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by edoakes — COMMENTED** [review]
This change won't solve the issue in isolation because the sender side (router) will still respect the limit.

The solution is also a little bit hacky to implicitly change the behavior like this -- I'd rather we do this at deployment time and update `max_ongoing_requests` then and log a warning.

**Comment by GeneDer** [comment]
@edoakes this is ready for another round of review. Changed to just log a warning and added a test for it

File: python/ray/serve/_private/config.py [hunk 1]
```diff
@@ -170,6 +171,51 @@ class DeploymentConfig(BaseModel):
...
+        """Helper to get a single max_batch_size form a method.
+
+        If the method doesn't uses batching, return 0.
+        """
+        if not hasattr(method, "_get_max_batch_size"):
+            return 0
+
+        return method._get_max_batch_size()
+
+    def check_max_batch_size_bounded(
```
Comment (edoakes) [inline python/ray/serve/_private/config.py:185]: I think it makes more sense to this inside of `serve.batching._BatchQueue` rather than adding this method parsing hackery. As written, this currently does not work if users use the `set_max_batch_size` setter on the method (which can be done dynamically at runtime in the replica).

We'll need access to `max_ongoing_requests` there, for which I'd suggest we add `deployment_config` as a private attribute of `ReplicaContext` (and set/update it properly in `replica.py` for that purpose). This will likely be generally useful in the future anyways.

Comment (GeneDer) [reply python/ray/serve/_private/config.py:185]: Oh cool, didn't know there is a way for batching to get `max_ongoing_requests`. Thanks for the suggestion Ed!

File: python/ray/serve/_private/replica.py [hunk 2]
```diff
@@ -636,6 +637,12 @@ async def reconfigure(
                     deployment_config.user_config
                 )
 
+            # We need to update internal replica context to reflect the new
+            # deployment_config and servable_object.
+            self._set_internal_replica_context(
+                servable_object=self._user_callable_wrapper.user_callable
+            )
+
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:645]: hmm why is this necessary?

Comment (edoakes) [reply python/ray/serve/_private/replica.py:645]: oh I see, this is also updating the deployment_config. The comment about `servable_object` is irrelevant here

File: python/ray/serve/batching.py [hunk 3]
```diff
@@ -112,6 +113,29 @@ def __init__(
             self._handle_batch_task = self._loop.create_task(
                 self._process_batches(handle_batch_func)
             )
+        self.check_max_batch_size_bounded()
+
+    def check_max_batch_size_bounded(self):
+        """Helper to check whether the max_batch_size is bounded.
+
+        Log a warning to configure `max_ongoing_requests` if it's bounded.
```
Comment (edoakes) [inline python/ray/serve/batching.py:121]:
```suggestion
    def warn_if_max_batch_size_exceeds_max_ongoing_requests(self):
        """Helper to check whether the max_batch_size is bounded.

        Log a warning to configure `max_ongoing_requests` if it's bounded.
```
unclear what "bounded" means without context.

Comment (edoakes) [reply python/ray/serve/batching.py:121]: this should be a private method as well

File: python/ray/serve/batching.py [hunk 4]
```diff
@@ -112,6 +113,29 @@ def __init__(
...
+        max_ongoing_requests = (
+            serve.get_replica_context()._deployment_config.max_ongoing_requests
+        )
+        if max_ongoing_requests < self.max_batch_size:
+            logger.warning(
+                f"`max_batch_size` ({self.max_batch_size}) is larger than "
+                f"`max_ongoing_requests` ({max_ongoing_requests}). The maximum "
+                f"ongoing request will be bounded by {max_ongoing_requests}. "
+                "To allow batching to reach the `max_batch_size` limits, please "
+                "configue `max_ongoing_requests` to be >= `max_batch_size`."
```
Comment (edoakes) [inline python/ray/serve/batching.py:132]:
```suggestion
                f"`max_batch_size` ({self.max_batch_size}) is larger than "
                f"`max_ongoing_requests` ({max_ongoing_requests}). This means "
                f"the replica will never receive a full batch. Please update "
                "`max_ongoing_requests` to be >= `max_batch_size."
```
try to be concise!

File: python/ray/serve/_private/replica.py (hunk 2 above)
Comment (GeneDer) [reply python/ray/serve/_private/replica.py:645]: Sg, will drop the comment about `servable_object`👍

**Review by edoakes — APPROVED** [review]

**Comment by GeneDer** [comment]
@edoakes this is ready for merge
